src/leo_cache_tran.erl

```diff
@@ -0,0 +1,198 @@
+         stop/0]).
+
+-export([tran/3,
+         has_tran/2,
+         done_tran/2]).
+
+-export([init/1,
+         handle_call/3,
+         handle_cast/2,
+         handle_info/2,
+         terminate/2,
+         code_change/3]).
+
+-define(PROCESSDB, leo_cache_tran_proc_db).
+-define(REPLYDB, leo_cache_tran_reply_db).
+
+
+%%--------------------------------------------------------------------
+%% API
+%%--------------------------------------------------------------------
+%% Function: start_link() -> {ok,Pid} | ignore | {error,Error}
+%% Description: Starts the server
+start_link() ->
+    gen_server:start_link({local, ?MODULE}, ?MODULE, [], []).
+
+stop() ->
+    gen_server:call(?MODULE, stop).
+
+%% @doc
+-spec(tran(pid(), atom(), binary()) ->
+             ok).
+tran(Pid, Tbl, Key) ->
+    gen_server:call(?MODULE, {tran, Pid, Tbl, Key}, ?TRAN_TIMEOUT).
+
+
+%% @doc
+-spec(has_tran(atom(), binary()) ->
+             {ok, any()} | {error, any()}).
+has_tran(Tbl, Key) ->
+    case catch gen_server:call(?MODULE, {has_tran, Tbl, Key}, ?TRAN_WAITTIME) of
+        {'EXIT', {timeout, _}} ->
+%%            gen_server:call(?MODULE, {unregister, Tbl, Key}, ?TRAN_TIMEOUT),
+            {error, timeout};
+        {'EXIT', Reason} ->
+%%            gen_server:call(?MODULE, {unregister, Tbl, Key}, ?TRAN_TIMEOUT),
+            {error, Reason};
+        Ret ->
+            Ret
+    end.
+
+%% @doc
+-spec(done_tran(atom(), binary())->
+            ok).
+done_tran(Tbl, Key) ->
+    gen_server:call(?MODULE, {done_tran, Tbl, Key}, ?TRAN_TIMEOUT).
+
+
+%%--------------------------------------------------------------------
+%% GEN_SERVER CALLBACKS
+%%--------------------------------------------------------------------
+%% Function: init(Args) -> {ok, State}          |
+%%                         {ok, State, Timeout} |
+%%                         ignore               |
+%%                         {stop, Reason}
+%% Description: Initiates the server
+init([]) ->
+    ets:new(?PROCESSDB, [named_table, set, private]),
+    ets:new(?REPLYDB, [named_table, set, private]),
+    {ok, unused, ?TRAN_TIMEOUT}.
+
+handle_call(stop, _From, State) ->
+    {stop, normal, ok, State};
+
+%%handle_call({unregister, Tbl, Key}, From, State) ->
+%%    {Pid, _Ref} = From,
+%%    case ets:lookup(?REPLYDB, {Tbl, Key}) of
+%%        [{{Tbl, Key}, ReplyList}] ->
+%%            ?debugVal(ReplyList),
+%%            ReplyList2 = lists:filter(fun({Pid2, _Ref2}) ->
+%%                                              Pid2 =/= Pid
+%%                                      end, ReplyList),
+%%            ?debugVal(ReplyList2),
+%%            ets:insert(?REPLYDB, {{Tbl, Key}, ReplyList2});
+%%        _ ->
+%%            void
+%%    end,
+%%    {reply, ok, State, ?TRAN_TIMEOUT};
+
+handle_call({tran, Pid, Tbl, Key}, _From, State) ->
+    MonitorRef = erlang:monitor(process, Pid),
+    ets:insert(?PROCESSDB, {MonitorRef, {Tbl, Key}}), 
+    ets:insert_new(?REPLYDB, {{Tbl, Key}, []}),
```

Comment:
Here is a "sub-optimal" solution to handle if two processes call `leo_cache_tran:tran/3`, simply both processes would be allowed to proceed to fetch the object, faster one would trigger the reply.

Alternatively, result of `ets:insert_new` can be checked and notify the caller. It would also need changes in `leo_gateway` to handle.